Add language context store.

starcke opened the pull request:
<!-- Thank you for submitting a pull request. Please read our pull request guidelines before
  submitting your pull request:
  https://github.com/github/vscode-codeql/blob/main/CONTRIBUTING.md#submitting-a-pull-request.
-->

This adds an explicit store to contain the Language filter state outside of the databases panel. It is based on `modeling-store.ts`.

There are two points that are somewhat interesting:
- We encapsulated the LanguageState inside the store, so that the users do not need to know how "All" is implemented. This might not be needed or sufficient in the future, but for now it seemed pretty nice to do.
- The `shouldInclude` method takes a string which is asymmetric with with the `setLanguageContext`. This is because we actually check with an arbitrary string (because that is what the database item has). I wonder if there is a better way to do this.

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

shati-patel reviewed (approved):
This looks good to me, though admittedly we paired, so happy to get another opinion too! 🙃 

(Also worth noting: we discussed adding tests for the language store, but since it doesn't do much on its own, we'll add tests when integrating with the individual components 🧪)

shati-patel commented on extensions/ql-vscode/src/language-context-store.ts:
```diff
@@ -0,0 +1,40 @@
+    this.onLanguageContextChangedEmitter.fire();
+  }
+
+  // TODO: comment on why string is used here
```
A leftover "TODO" 👀 Do you want to add a comment here, or shall we remove this? 😄

starcke replied:
I was leaving it to highlight the question, but I did that in the PR message, so maybe I can just remove it :)

starcke commented:
I removed the TODO (and added a comment). Then I fixed the conflicts.

Lastly I noticed that when the extension was initially started up the "All" value didnt show as selected, because the context value wasnt set yet.
The last commit fixes this (by making the context look for empty string instead) and moves the context manipulation into the state.

charisk commented on extensions/ql-vscode/src/language-context-store.ts:
```diff
@@ -0,0 +1,53 @@
+  // where the language is not always a known one.
+  // The semantics of such an unknown langauge is that it is
+  // only included if the current language context is "All".
+  public shouldInclude(language: string): boolean {
```
Can we use [`isQueryLanguage`](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/src/common/query-language.ts#L62) here to help us out?

starcke replied:
Not sure what we would use it for. We could check that the string is a query language, but because the only thing that can be put into the state is a query language or all, then the check inside the second clause of the disjunction should amount to the same thing. But I might be missing something?

charisk replied:
We've made the argument of this function a `string` because the caller could have a `string` rather than a `QueryLanguage` (if I'm understanding this right). But only valid QueryLanguage values should be allowed. So it'd be good for the argument to change to a `QueryLanguage` and the caller to user `isQueryLanguage` to make sure it's sending the right thing.

starcke replied:
The trouble with having the caller use `isQueryLanguage` is that makes it hard to preserve the semantics (that unknown values should be shown when "All" is selected) without leaking "All" outside the store, or having to expose a method like "IsAllSelected".

One alternative is to pass `QueryLanguage | undefined` (because I actually need that for the queries panel). However that does make it more annoying for the callers, as they then have to convert their `string` to a `QueryLanguge | undefined`. But that is not the end of the world, and we could do that as a helper method similar to `isQueryLanguage`.

The last alternative is to expose `LanguageFilter` outside the store, which is also fine but it seemed nice to encapsulate the "All" implementation.

charisk replied:
`QueryLanguage | undefined` sounds like it makes the most sense to me (and I'm okay with the callers having to put some extra effort in) but I'm also happy with whatever you decide - I've not thought about this deeply, and it's easily changeable if we find that it's causing problems too.

starcke replied:
Pushed a new version that takes in `QueryLanguage | undefined`.

charisk commented on extensions/ql-vscode/src/extension.ts:
```diff
@@ -774,10 +775,15 @@ async function activateWithInstalledDistribution(
   ctx.subscriptions.push(dbm);
+
+  void extLogger.log("Initializing language context.");
+  const languageContext = new LanguageContextStore(app);
```
Are these changes meant to be behind a feature flag or did I imagine that?

starcke replied:
I thought they were, but I cannot see one before this PR either?

shati-patel replied:
The "Languages" dropdown is feature-flagged in the UI here: https://github.com/github/vscode-codeql/blob/6aab4b4090600ce4d6ebd9183feb2cd215b809ac/extensions/ql-vscode/package.json#L1080

So it might make sense to feature-flag this code too 🎏

starcke replied:
:+1: I have had the canary flag as a "standard" flag, so I didnt even think about it :doh:

How do we normally flag this? If we just omit the creation of the languageContext then we'll have a lot of places where it is suddenly undefined, which seems more risky than having it in the first place?

shati-patel replied:
For the queries panel, we check for the flag before initializing: https://github.com/github/vscode-codeql/blob/2e45dcd7b9a0aa01deeb1605e18d8eaa6515c14e/extensions/ql-vscode/src/queries-panel/queries-module.ts#L29-L33 

But in this case, I think it's fairly harmless to create it without feature-flagging 🤔 I don't really know what's "best" or "safest" 😅

charisk replied:
Ideally no code that does language filtering stuff should be executed if the flag is off. But I think it's okay to be lenient - it's all about us having enough confidence when releasing rather than sticking to specific rules.

starcke replied:
I dont see a good way of feature-flagging this. But happy to discuss more sync.

charisk commented on extensions/ql-vscode/src/language-context-store.ts:
```diff
@@ -0,0 +1,53 @@
+  public readonly onLanguageContextChanged: AppEvent<void>;
+  private readonly onLanguageContextChangedEmitter: AppEventEmitter<void>;
+
+  private state: LanguageFilter;
```
(Optional) I'd personally call this `languageFilter` rather than `state` to make code further down easier to read.

starcke commented:
Are you both happy with the changes?

charisk reviewed (approved):
🚢
